toolkit/tools/internal/isogenerator/isogenerator.go

```diff
@@ -0,0 +1,338 @@
+	// - vmlinuz
+	// - isolinux.bin (for BIOS boot)
+	// - boot.cat (for BIOS boot)
+	IsoOsFilesDirPath string
+	// The path of the initrd file to use in the ISO.
+	InitrdPath string
+}
+
+type isoGenInfo struct {
+	Config         IsoGenConfig
+	EfiBootImgPath string
+}
+
+func GenerateIso(config IsoGenConfig) error {
+	err := os.MkdirAll(config.StagingDirPath, os.ModePerm)
+	if err != nil {
+		return fmt.Errorf("failed to create ISO staging directory (%s):\n%w", config.StagingDirPath, err)
+	}
+
+	efiBootImgPath := filepath.Join(config.StagingDirPath, efiBootImgPathRelativeToIsoRoot)
+
+	info := isoGenInfo{
+		Config:         config,
+		EfiBootImgPath: efiBootImgPath,
+	}
+
+	err = prepareIsoBootLoaderFilesAndFolders(info)
+	if err != nil {
+		return err
+	}
+
+	err = buildIsoImage(info)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func buildIsoImage(info isoGenInfo) error {
+	logger.Log.Infof("Generating ISO image under '%s'.", info.Config.OutputFilePath)
+
+	// For detailed parameter explanation see: https://linux.die.net/man/8/mkisofs.
+	// Mkisofs requires all argument paths to be relative to the input directory.
+	mkisofsArgs := []string{}
+
+	mkisofsArgs = append(mkisofsArgs,
+		// General mkisofs parameters.
+		"-R", "-l", "-D", "-J", "-joliet-long", "-o", info.Config.OutputFilePath, "-V", DefaultVolumeId)
+
+	if info.Config.EnableBiosBoot {
+		mkisofsArgs = append(mkisofsArgs,
+			// BIOS bootloader, params suggested by https://wiki.syslinux.org/wiki/index.php?title=ISOLINUX.
+			"-b", filepath.Join(info.Config.IsoOsFilesDirPath, "isolinux.bin"),
+			"-c", filepath.Join(info.Config.IsoOsFilesDirPath, "boot.cat"),
+			"-no-emul-boot", "-boot-load-size", "4", "-boot-info-table")
+	}
+
+	mkisofsArgs = append(mkisofsArgs,
+		// UEFI bootloader.
+		"-eltorito-alt-boot", "-e", efiBootImgPathRelativeToIsoRoot, "-no-emul-boot",
+
+		// Directory to convert to an ISO.
+		info.Config.StagingDirPath)
+
+	// Note: mkisofs has a noisy stderr.
+	err := shell.ExecuteLive(true /*squashErrors*/, "mkisofs", mkisofsArgs...)
+	if err != nil {
+		return fmt.Errorf("failed to generate ISO using mkisofs:\n%w", err)
+	}
+
+	return nil
+}
+
+// prepareIsoBootLoaderFilesAndFolders copies the files required by the ISO's bootloader
+func prepareIsoBootLoaderFilesAndFolders(info isoGenInfo) (err error) {
+	err = setUpIsoGrub2Bootloader(info)
+	if err != nil {
+		return err
+	}
+
+	err = createVmlinuzImage(info)
+	if err != nil {
+		return err
+	}
+
+	err = copyInitrd(info)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// copyInitrd copies a pre-built initrd into the isolinux folder.
+func copyInitrd(info isoGenInfo) error {
+	initrdDestinationPath := filepath.Join(info.Config.StagingDirPath, info.Config.IsoOsFilesDirPath, "initrd.img")
+
+	logger.Log.Debugf("Copying initrd from '%s'.", info.Config.InitrdPath)
+
+	return file.Copy(info.Config.InitrdPath, initrdDestinationPath)
+}
+
+func setUpIsoGrub2Bootloader(info isoGenInfo) (err error) {
```

Comment:
nit: maybe 'createIsoBootloaderImage()'?

Reply:
`setUpIsoGrub2Bootloader` is the name used by `IsoMaker`. I'd rather keep the name the same to make the history clearer to follow.